pep-0695.rst

```diff
@@ -407,30 +361,254 @@ At runtime, a ``type`` statement will generate an instance of
+::
+
+    T = 0
+
+    @decorator(T)  # Argument expression `T` evaluates to 0
+    class ClassA[T](Sequence[T]):
+        T = 1
+
+        # All methods below should result in a type checker error
+        # "type parameter 'T' already in use" because they are using the
+        # type parameter 'T', which is already in use by the outer scope
+        # 'ClassA'.
+        def method1[T](self):
+            ...
+
+        def method2[T](self, x = T):  # Parameter 'x' gets default value of 1
+            ...
+
+        def method3[T](self, x: T):  # Parameter 'x' has type T (scoped to method3)
+            ...
+
+
+Symbols referenced in inner scopes are resolved using existing rules except
+that type parameter scopes are also considered during name resolution.
+
+::
+    T = 0
+
+    # T refers to the global variable
+    print(T)  # Prints 0
+
+    class Outer[T]:
+        T = 1
+
+        # T refers to the local variable scoped to class 'Outer'
+        print(T)  # Prints 1
+
+        class Inner1:
+            T = 2
+
+            # T refers to the local type variable within 'Inner1'
+            print(T)  # Prints 2
+
+            def inner_method(self):
+                # T refers to the type parameter scoped to class 'Outer';
+                # If 'Outer' did not use the new type parameter syntax,
+                # this would instead refer to the global variable 'T'
+                print(T)  # Prints 'T'
+
+        def outer_method(self):
+            T = 3
+
+            # T refers to the local variable within 'outer_method'
+            print(T)  # Prints 3
+
+            def inner_func():
+                # T refers to the variable captured from 'outer_method'
+                print(T)  # Prints 3
+
+
+When the new type parameter syntax is used for a generic class, assignment
+expressions are not allowed within the argument list for the class definition.
+Likewise, with functions that use the new type parameter syntax, assignment
+expressions are not allowed within parameter or return type annotations, nor
+are they allowed within the expression that defines a type alias.
+
+This restriction is necessary because expressions evaluated within the
+new lexical scope should not introduce symbols within that scope other than
+the defined type parameters.
+
+::
+
+    class ClassA[T]((x := Sequence[T])): ...  # Syntax error: assignment expression not allowed
+
+    def func1[T](val: (x := int)): ...  # Syntax error: assignment expression not allowed
+
+    def func2[T]() -> (x := Sequence[T]): ...  # Syntax error: assignment expression not allowed
+
+    type Alias1[T] = (x := list[T])  # Syntax error: assignment expression not allowed
+
+
+Accessing Type Parameters at Runtime
+------------------------------------
+
+A new read-only attribute called ``__type_variables__`` is available on class,
+function, and type alias objects. This attribute is a tuple of the active
+type variables that are visible within the scope of that class, function,
```

Comment:
To be clear the tuple contains the `TypeVar` objects created on for the given class, function or type alias, right?

Reply:
No, this tuple contains all of the active type variables including those created on the given class, function or type alias and any outer scopes that have defined type variables. This is important for runtime type evaluation (e.g. typing.get_type_hints). 

There is a separate tuple called `__parameters__` that contains only the type variables associated with the class, function or type alias.

Reply:
Ah, the text actually says so, using the term "active".

My question was actually, does the tuple contain strings or `TypeVar` instances?

Reply:
It contains `TypeVar` instances. If you want to access the names, you can ask each `TypeVar` object for its own name.

Reply:
Thanks, I added a sentence at the end of the paragraph to clarify this.